lib/rspec-puppet/cache.rb

```diff
@@ -11,10 +11,10 @@ def initialize(&default_proc)
     end
 
     def get(*args, &blk)
+      expire!
       if !@cache.has_key? args
         @cache[args] = (blk || @default_proc).call(*args)
         @lra << args
-        expire!
```

Comment:
[1,2,3,4], becomes [1,2,3,4,1] when appending 1. if you do a unique, the result is [1,2,3,4]. if instead reversing the array first it becomes [1,4,3,2,1] and a unique then gives [1,4,3,2], which reversed (again) is [2,3,4,1]. The 1 was "touched" and moved last in the array. Now if cache max was 4, and we add an element 5, we first get [2,3,4,1,5], and the expire will evict the first 3 and we end up with [1, 5]. The '1' survived because it was used later in time than 2, 3, and 4.

That is the idea at least.

There are other ways of doing the same thing - in pseudo code:
if cache-hit
  find the corresponding entry in @ira and move it last in @ira (i.e. this indicates "used last")
```
if cache-miss
   evict from cache if there are too many entries, but only evict one (seems bad to evict all of them)
  add the args to @ira, and @cache
end
```

However that is not the problem here (unless a very odd combination of args/cache is interacting). The above is however a better cache implementation if the idea is to just cache catalogs given its args (input); hard to tell if there are other requirements.

Will continue looking at the other problem.